BWIndex.py

```diff
@@ -0,0 +1,239 @@
+from graphviz import Digraph
+import sys, argparse
+
+
+class BWIndex():
+	def __init__(self, genome=None, genomefile=None):
+		if genome is not None:
+			self.Text = genome
+		elif genomefile is not None:
+			self.loadGenomeFromFile(genomefile)
+		else:
+			print("You must supply a genome file or a genome string.")
+			sys.exit()
+		if self.Text[-1] != '$':
+			print("Genome must be terminated by $")
+			sys.exit()
+		self.buildTree()
+		self.Solve()
+		self.Count = dict()
+		self.letters = {i for i in self.BWT}
+		self.alphabet = [i for i in self.letters]
+		self.alphabet.sort()
+		self.computeCountArray()
+
+		self.FirstOccurrence = dict()
+		freq = dict()
+		for i in self.BWT:
+			freq[i] = freq.get(i, 0) + 1
+		index = 0
+		self.FirstOccurrence['$'] = 0
+		for j in range(1, len(self.alphabet)):
+			prevLetter = self.alphabet[j - 1]
+			letter = self.alphabet[j]
+			index += freq[prevLetter]
+			self.FirstOccurrence[letter] = index
+
+	def buildTree(self):
+		self.G = {0: []}  # parent to child, int -> list
+		self.H = dict()  # child to parent, int -> int
+		self.nodeCount = 1
+		self.NS = dict()  # node start index
+		self.NE = dict()  # node end index
+		self.SS = dict()  # suffix start index
+		for ix in range(len(self.Text)):
+			# print(ix)
+			self.threadSuffix(ix)
+
+	def exportIndex(self, filename, SA_gap=None, C_gap=None):
+		g = open(filename, 'w')
+		self.SA_gap = SA_gap
+		self.C_gap = C_gap
+		g.write(self.BWT + '\n')
+		if SA_gap is None:
+			g.write(','.join(str(i) + ';' + str(s) for i, s in enumerate(self.suffixArray)) + '\n')
+		else:
+			g.write(','.join(str(i) + ';' + str(s) for i, s in enumerate(self.suffixArray) if s % SA_gap == 0) + '\n')
+		g.write(','.join(str(x) for x in sorted(self.Count.keys())) + '\n')
+		for x in sorted(self.Count.keys()):
+			if C_gap is None:
+				g.write(','.join(str(self.Count[x][i]) for i in range(len(self.Count[x]))) + '\n')
+			else:
+				g.write(','.join(str(self.Count[x][i]) for i in range(len(self.Count[x])) if i % C_gap == 0) + '\n')
+		g.write(','.join(str(self.FirstOccurrence[x]) for x in sorted(self.FirstOccurrence.keys())) + '\n')
+		g.write(str(SA_gap) + '\n')
+		g.write(str(C_gap) + '\n')
+
+	def Solve(self):
+		S = [('', 0)]
+		transform = ''
+		suffixArray = []
+		while S:
+			node = S.pop()
+			nodeIx = node[1]
+			if self.G.get(nodeIx) is not None:
+				if self.G[nodeIx] != []:
+					children = [(self.Text[self.NS[child]], child) for child in self.G[nodeIx]]
+					children.sort(reverse=True)
+					S += children
+			else:
+				transform += self.Text[self.SS[nodeIx] - 1]

```

Comment:
Is transform going to end up being a huge string? If so, using `+=` here will create a new string copy every time you invoke it. A better strategy for building large strings is to add them all to a list and then once the list is complete call `"".join` on the list to construct a string just once.